src/main/java/com/google/sps/servlets/TripsServlet.java

```diff
@@ -86,48 +87,111 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   }
 
   @Override
-  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("application/json;");
+  public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
     UserService userService = UserServiceFactory.getUserService();
-    if (userService.isUserLoggedIn()) {
-      String userEmail = userService.getCurrentUser().getEmail();
-      long timestamp = System.currentTimeMillis();
-      
-      String requestBody = request.getReader().lines().collect(Collectors.joining(System.lineSeparator()));
-      JsonObject jsonObject = JsonParser.parseString(requestBody).getAsJsonObject();
-      
-      Trip trip = Trip.builder()
-                    .setTitle(jsonObject.getAsJsonPrimitive(Trip.ENTITY_PROPERTY_TITLE).getAsString())
-                    .setHotelID(jsonObject.getAsJsonPrimitive(Trip.ENTITY_PROPERTY_HOTEL_ID).getAsString())
-                    .setHotelName(jsonObject.getAsJsonPrimitive(Trip.ENTITY_PROPERTY_HOTEL_NAME).getAsString())
-                    .setHotelImage(jsonObject.getAsJsonPrimitive(Trip.ENTITY_PROPERTY_HOTEL_IMAGE).getAsString())
-                    .setRating(jsonObject.getAsJsonPrimitive(Trip.ENTITY_PROPERTY_RATING).getAsDouble())
-                    .setOwner(userEmail)
-                    .setTimestamp(timestamp)
-                    .build();
-      Entity tripEntity = convertTripToEntity(trip);
+    if (!userService.isUserLoggedIn()) {
+      response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+      return;
+    }
+    String userEmail = userService.getCurrentUser().getEmail();
+    
+    String requestBody = request.getReader().lines().collect(Collectors.joining(System.lineSeparator()));
+    JsonObject jsonObject = JsonParser.parseString(requestBody).getAsJsonObject();
 
-      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-      datastore.put(tripEntity);
+    // true if the client is attempting to do an update, false if insert
+    boolean isUpdating = jsonObject.has("timestamp");
+    long timestamp = isUpdating ? jsonObject.getAsJsonPrimitive(Trip.ENTITY_PROPERTY_TIMESTAMP).getAsLong() : System.currentTimeMillis();
```

Comment:
The create flow doesn't look like idempotent. Maybe you can keep the post method, and refactor everything under a shared private method?

Reply:
sg. I think the update functionality is idempotent but since the endpoint takes different JSON inputs it wouldn't truly be an idiomatic use of `PUT`.

Reply:
Done, refactored a lot of stuff into helpers/a util class.